In this notebook, we are going to prepare our data for prediction of Remaining Service Life, The approach has been included in the project report


We have 4 datasets, data collected from 2014-2017, we are going to join them and generate a time series dataset. Additionally we will attempt to add some weather and traffic data based on the geolocational data provided and also drop some data we might not need as it does not directly influence the PCI value.

# Approach
We will import one dataset, of any one year and make functions for that ,then write general code for preparing all datasets then combine them later on.

In [5]:
import pandas as pd

In [6]:
d2015=pd.read_excel("G:\ml notebooks\Road_RUL\datasets\datasets\historical_data\open_data_submission-_2015_pavement_final_0.xlsx")

d2015

Section ID Highway   DIR     LHRS  Offset  FROM_Distance  TO_Distance  \
0            NaN      NaN  NaN      NaN     NaN            NaN          NaN   
1            NaN      QEW    E  10004.0     0.0          0.230        4.658   
2            NaN      QEW    W  10004.0     0.0          0.230        4.658   
3            NaN      QEW    E  10014.0     0.0          4.658       13.227   
4            NaN      QEW    W  10014.0     0.0          4.658       13.227   
...          ...      ...  ...      ...     ...            ...          ...   
1825         NaN     1000    B  99998.0     0.0          0.000        8.600   
1826         NaN     1000    B  99998.0     8.6          8.600       20.300   
1827         NaN     1000    B  99998.0    20.3         20.300       23.200   
1828         NaN     1000    B  99998.0    23.2         23.200       27.500   
1829         NaN     1000    B  99998.0    27.5         27.500       37.300   

        PCI   DMI   IRI Pave_Type                     Pavement_Section_From  \
0       NaN   NaN   NaN       NaN                                       NaN   
1     85.00  8.89  1.17        AC                 FORT ERIE - CENTRAL AV IC   
2     85.13  8.83  1.08        AC                 FORT ERIE - CENTRAL AV IC   
3     80.84  8.18  0.87        AC                            GILMORE RD I/C   
4     79.18  8.03  0.91        AC                            GILMORE RD I/C   
...     ...   ...   ...       ...                                       ...   
1825  85.01  8.65  0.88        AC                         WEST JCT OF HWY 2   
1826  79.14  8.15  1.06        AC                          ROSE ISLAND LANE   
1827  79.49  8.18  1.05        AC                   2.8 KM EAST OF ROCKPORT   
1828  72.33  7.91  1.68        AC  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829  78.93  8.16  1.10        AC                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To FUNC_CLASS  
0                                            NaN        NaN  
1                    FORT ERIE - GILMORE RD IC-5        FWY  
2                    FORT ERIE - GILMORE RD IC-5        FWY  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12        FWY  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12        FWY  
...                                          ...        ...  
1825                            ROSE ISLAND LANE        COL  
1826                             NARROWS LANE RD        COL  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)        COL  
1828                  0.5 KM W OF MALLORYTOWN RD        COL  
1829                  EAST JUNCTION OF HIGHWAY 2        COL  

[1830 rows x 14 columns]

In [7]:
d2015.columns

Index(['Section ID', 'Highway ', 'DIR', 'LHRS', 'Offset', 'FROM_Distance',
       'TO_Distance', 'PCI', 'DMI', 'IRI', 'Pave_Type',
       'Pavement_Section_From', 'Pavement_Section_To', 'FUNC_CLASS'],
      dtype='object')

We definitely do not need the section ID, Higway, LHRS, Offset,From and To distance, Pave Type,Func class as :
1. They do not influence PCI DMI or IRI value directly or even indirectly.
2. It is out of our scope here to consider those as well as we cannot get that data from our image inputs


In [8]:
d2015.drop(["Section ID","Highway ","LHRS","Offset","FROM_Distance","TO_Distance","Pave_Type","FUNC_CLASS"],axis=1,inplace=True)
d2015

DIR    PCI   DMI   IRI                     Pavement_Section_From  \
0     NaN    NaN   NaN   NaN                                       NaN   
1       E  85.00  8.89  1.17                 FORT ERIE - CENTRAL AV IC   
2       W  85.13  8.83  1.08                 FORT ERIE - CENTRAL AV IC   
3       E  80.84  8.18  0.87                            GILMORE RD I/C   
4       W  79.18  8.03  0.91                            GILMORE RD I/C   
...   ...    ...   ...   ...                                       ...   
1825    B  85.01  8.65  0.88                         WEST JCT OF HWY 2   
1826    B  79.14  8.15  1.06                          ROSE ISLAND LANE   
1827    B  79.49  8.18  1.05                   2.8 KM EAST OF ROCKPORT   
1828    B  72.33  7.91  1.68  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829    B  78.93  8.16  1.10                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To  
0                                            NaN  
1                    FORT ERIE - GILMORE RD IC-5  
2                    FORT ERIE - GILMORE RD IC-5  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
...                                          ...  
1825                            ROSE ISLAND LANE  
1826                             NARROWS LANE RD  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  
1828                  0.5 KM W OF MALLORYTOWN RD  
1829                  EAST JUNCTION OF HIGHWAY 2  

[1830 rows x 6 columns]

In [9]:
#Remove first low as all fields are empty
d2015.drop(index=0,inplace=True)


In [10]:
d2015

DIR    PCI   DMI   IRI                     Pavement_Section_From  \
1      E  85.00  8.89  1.17                 FORT ERIE - CENTRAL AV IC   
2      W  85.13  8.83  1.08                 FORT ERIE - CENTRAL AV IC   
3      E  80.84  8.18  0.87                            GILMORE RD I/C   
4      W  79.18  8.03  0.91                            GILMORE RD I/C   
5      N  81.24  8.34  1.01               1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                       ...   
1825   B  85.01  8.65  0.88                         WEST JCT OF HWY 2   
1826   B  79.14  8.15  1.06                          ROSE ISLAND LANE   
1827   B  79.49  8.18  1.05                   2.8 KM EAST OF ROCKPORT   
1828   B  72.33  7.91  1.68  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829   B  78.93  8.16  1.10                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To  
1                    FORT ERIE - GILMORE RD IC-5  
2                    FORT ERIE - GILMORE RD IC-5  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
5                 NIAG FALLS - LYONS CK RD IC-21  
...                                          ...  
1825                            ROSE ISLAND LANE  
1826                             NARROWS LANE RD  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  
1828                  0.5 KM W OF MALLORYTOWN RD  
1829                  EAST JUNCTION OF HIGHWAY 2  

[1829 rows x 6 columns]

So we have deleted columns we do not need

In [11]:
d2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1829 entries, 1 to 1829
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   DIR                    1829 non-null   object 
 1   PCI                    1829 non-null   float64
 2   DMI                    1829 non-null   float64
 3   IRI                    1829 non-null   float64
 4   Pavement_Section_From  1829 non-null   object 
 5   Pavement_Section_To    1829 non-null   object 
dtypes: float64(3), object(3)
memory usage: 85.9+ KB


In [12]:
d2015.isna().sum(axis=0)

DIR                      0
PCI                      0
DMI                      0
IRI                      0
Pavement_Section_From    0
Pavement_Section_To      0
dtype: int64

There are no null values

In [13]:
d2015.describe()

PCI          DMI          IRI
count  1829.000000  1829.000000  1829.000000
mean     76.432991     8.225921     1.663013
std      11.814475     0.937294     0.950915
min      24.550000     0.000000     0.000000
25%      71.230000     7.900000     1.000000
50%      78.750000     8.240000     1.320000
75%      84.030000     8.830000     2.130000
max      99.170000    10.000000     6.530000

We see that the average PCI is 76 and the max is 99.17, which is basically a perfect road.


In [14]:
d2015

DIR    PCI   DMI   IRI                     Pavement_Section_From  \
1      E  85.00  8.89  1.17                 FORT ERIE - CENTRAL AV IC   
2      W  85.13  8.83  1.08                 FORT ERIE - CENTRAL AV IC   
3      E  80.84  8.18  0.87                            GILMORE RD I/C   
4      W  79.18  8.03  0.91                            GILMORE RD I/C   
5      N  81.24  8.34  1.01               1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                       ...   
1825   B  85.01  8.65  0.88                         WEST JCT OF HWY 2   
1826   B  79.14  8.15  1.06                          ROSE ISLAND LANE   
1827   B  79.49  8.18  1.05                   2.8 KM EAST OF ROCKPORT   
1828   B  72.33  7.91  1.68  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829   B  78.93  8.16  1.10                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To  
1                    FORT ERIE - GILMORE RD IC-5  
2                    FORT ERIE - GILMORE RD IC-5  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
5                 NIAG FALLS - LYONS CK RD IC-21  
...                                          ...  
1825                            ROSE ISLAND LANE  
1826                             NARROWS LANE RD  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  
1828                  0.5 KM W OF MALLORYTOWN RD  
1829                  EAST JUNCTION OF HIGHWAY 2  

[1829 rows x 6 columns]

## We now try to get weather and traffic data based on the Pavement Section From and To columns which can give us an approximate location


In [15]:
addresses=d2015[["Pavement_Section_From","Pavement_Section_To"]].copy()

In [16]:
addresses

Pavement_Section_From  \
1                    FORT ERIE - CENTRAL AV IC   
2                    FORT ERIE - CENTRAL AV IC   
3                               GILMORE RD I/C   
4                               GILMORE RD I/C   
5                  1.0 KM W OF NETHERBY RD I/C   
...                                        ...   
1825                         WEST JCT OF HWY 2   
1826                          ROSE ISLAND LANE   
1827                   2.8 KM EAST OF ROCKPORT   
1828  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To  
1                    FORT ERIE - GILMORE RD IC-5  
2                    FORT ERIE - GILMORE RD IC-5  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
5                 NIAG FALLS - LYONS CK RD IC-21  
...                                          ...  
1825                            ROSE ISLAND LANE  
1826                             NARROWS LANE RD  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  
1828                  0.5 KM W OF MALLORYTOWN RD  
1829                  EAST JUNCTION OF HIGHWAY 2  

[1829 rows x 2 columns]

Now we have a list of addresses from which we can get the latitude and longitude.

In [17]:
from dotenv import load_dotenv
import os
def configure():
    load_dotenv()
    

In [18]:
configure()
api_key_bing=os.getenv('api_key1')

In [19]:
import geocoder
import geopy

In [20]:
geolocator=geopy.geocoders.Bing(api_key_bing)

In [21]:
addresses

Pavement_Section_From  \
1                    FORT ERIE - CENTRAL AV IC   
2                    FORT ERIE - CENTRAL AV IC   
3                               GILMORE RD I/C   
4                               GILMORE RD I/C   
5                  1.0 KM W OF NETHERBY RD I/C   
...                                        ...   
1825                         WEST JCT OF HWY 2   
1826                          ROSE ISLAND LANE   
1827                   2.8 KM EAST OF ROCKPORT   
1828  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To  
1                    FORT ERIE - GILMORE RD IC-5  
2                    FORT ERIE - GILMORE RD IC-5  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
5                 NIAG FALLS - LYONS CK RD IC-21  
...                                          ...  
1825                            ROSE ISLAND LANE  
1826                             NARROWS LANE RD  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  
1828                  0.5 KM W OF MALLORYTOWN RD  
1829                  EAST JUNCTION OF HIGHWAY 2  

[1829 rows x 2 columns]

In [22]:
def latlong(addressdata):
    location=geolocator.geocode(addressdata)
    if hasattr(location,'latitude') and hasattr(location,'longitude'):
        return(location.latitude,location.longitude)
    else:
        return 0,0

addresses["Lat"]=0
addresses["Lon"]=0
for ind in addresses.index:
    gg=latlong(addresses['Pavement_Section_From'][ind])
    addresses["Lat"][ind],addresses['Lon'][ind]=gg
    

C:\Users\NAVEEN\AppData\Local\Temp\ipykernel_2148\179802196.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  addresses["Lat"][ind],addresses['Lon'][ind]=gg
C:\Users\NAVEEN\AppData\Local\Temp\ipykernel_2148\179802196.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  addresses["Lat"][ind],addresses['Lon'][ind]=gg


In [23]:
addresses['Lon'].value_counts()[0]
#Just two occurences where Latitude and Longitude could not be found, so we can just remove those occurences

2

In [24]:
addresses=addresses[addresses['Lat']!=0]

In [25]:
if 0 in addresses['Lon']:
    print("YES")
else:
    print("NO")

NO


In [26]:
addresses

Pavement_Section_From  \
1                    FORT ERIE - CENTRAL AV IC   
2                    FORT ERIE - CENTRAL AV IC   
3                               GILMORE RD I/C   
4                               GILMORE RD I/C   
5                  1.0 KM W OF NETHERBY RD I/C   
...                                        ...   
1825                         WEST JCT OF HWY 2   
1826                          ROSE ISLAND LANE   
1827                   2.8 KM EAST OF ROCKPORT   
1828  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To        Lat         Lon  
1                    FORT ERIE - GILMORE RD IC-5  42.929756  -78.918900  
2                    FORT ERIE - GILMORE RD IC-5  42.929756  -78.918900  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  34.899558  -81.064318  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  34.899558  -81.064318  
5                 NIAG FALLS - LYONS CK RD IC-21 -34.971001  138.624146  
...                                          ...        ...         ...  
1825                            ROSE ISLAND LANE  44.251579  -72.571407  
1826                             NARROWS LANE RD  51.718254   -1.238986  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  -6.471565  106.849001  
1828                  0.5 KM W OF MALLORYTOWN RD  28.041248  -97.042511  
1829                  EAST JUNCTION OF HIGHWAY 2  44.456743  -75.861192  

[1827 rows x 4 columns]

Since we have used Bing Geocoder, the resutls are not all accurate, but in one of our datasets, i.e the data collected in 2014, latitude and longitude has been provided, so for accuracy, we can just use those and some values are not given, for those, we can use the above values.

Zero values have been removed, now we get the weather data using OpenMeteo API

In [27]:
d2014=pd.read_csv("G:\ml notebooks\Road_RUL\datasets\datasets\historical_data\mto_pavement_condition_indices_2014.csv")

In [28]:
d2014

SectionID Highway  DIR     LHRS  Offset  FROM_Distance  TO_Distance  \
0           NaN     NaN  NaN      NaN     NaN            NaN          NaN   
1      100010.0     QEW    E  10004.0     0.0          0.230        4.658   
2      100020.0     QEW    W  10004.0     0.0          0.230        4.658   
3      100030.0     QEW    E  10014.0     0.0          4.658       13.227   
4      100040.0     QEW    W  10014.0     0.0          4.658       13.227   
...         ...     ...  ...      ...     ...            ...          ...   
1814   118510.0    7296    B  85600.0     0.0          0.000        2.610   
1815   118520.0    7297    B  85610.0     0.0          0.000        0.930   
1816   118530.0    7298    B  85620.0     0.0          0.000        2.000   
1817   118540.0    7909    B  98100.0     0.0          0.000        9.960   
1818   118550.0    7910    B  98115.0     0.0          0.000        1.600   

        PCI   DMI   IRI Pave_Type                       Pavement_Section_From  \
0       NaN   NaN   NaN       NaN                                         NaN   
1     85.74  8.97  1.17        AC                 FORT ERIE - GILMORE RD IC-5   
2     91.51  9.54  1.08        AC                   FORT ERIE - CENTRAL AV IC   
3     89.05  9.09  0.87        AC  NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12   
4     88.16  9.03  0.91        AC                              GILMORE RD I/C   
...     ...   ...   ...       ...                                         ...   
1814  79.81  8.65  1.57        AC                                   ROBINS RD   
1815  92.02  9.83  1.36        AC                                      OKE DR   
1816  91.95  9.63  1.13        AC                               ONTARIO ST IC   
1817  87.60  8.93  0.87        AC                              LAKE FOREST DR   
1818  90.34  9.51  1.20        AC                                  HWY 400 IC   

                             Pavement_Section_To   Latitude  Longitude  \
0                                            NaN        NaN        NaN   
1                      FORT ERIE - CENTRAL AV IC  42.919138 -78.968180   
2                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073   
3                                 GILMORE RD I/C  42.975538 -79.038586   
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961   
...                                          ...        ...        ...   
1814                                  END OF HWY        NaN        NaN   
1815                                  END OF HWY        NaN        NaN   
1816                                  END OF HWY        NaN        NaN   
1817                                     HWY 559        NaN        NaN   
1818                         HWY 7909 (NOBEL RD)        NaN        NaN   

     FUNC_CLASS  
0           NaN  
1           FWY  
2           FWY  
3           FWY  
4           FWY  
...         ...  
1814        LOC  
1815        LOC  
1816        LOC  
1817        LOC  
1818        LOC  

[1819 rows x 16 columns]

In [29]:
d2014.drop(["SectionID","Highway","LHRS","Offset","FROM_Distance","TO_Distance","Pave_Type","FUNC_CLASS"],axis=1,inplace=True)
d2014

DIR    PCI   DMI   IRI                       Pavement_Section_From  \
0     NaN    NaN   NaN   NaN                                         NaN   
1       E  85.74  8.97  1.17                 FORT ERIE - GILMORE RD IC-5   
2       W  91.51  9.54  1.08                   FORT ERIE - CENTRAL AV IC   
3       E  89.05  9.09  0.87  NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12   
4       W  88.16  9.03  0.91                              GILMORE RD I/C   
...   ...    ...   ...   ...                                         ...   
1814    B  79.81  8.65  1.57                                   ROBINS RD   
1815    B  92.02  9.83  1.36                                      OKE DR   
1816    B  91.95  9.63  1.13                               ONTARIO ST IC   
1817    B  87.60  8.93  0.87                              LAKE FOREST DR   
1818    B  90.34  9.51  1.20                                  HWY 400 IC   

                             Pavement_Section_To   Latitude  Longitude  
0                                            NaN        NaN        NaN  
1                      FORT ERIE - CENTRAL AV IC  42.919138 -78.968180  
2                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
3                                 GILMORE RD I/C  42.975538 -79.038586  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
...                                          ...        ...        ...  
1814                                  END OF HWY        NaN        NaN  
1815                                  END OF HWY        NaN        NaN  
1816                                  END OF HWY        NaN        NaN  
1817                                     HWY 559        NaN        NaN  
1818                         HWY 7909 (NOBEL RD)        NaN        NaN  

[1819 rows x 8 columns]

In [30]:
#Remove first low as all fields are empty
d2014.drop(index=0,inplace=True)

In [31]:
d2015

DIR    PCI   DMI   IRI                     Pavement_Section_From  \
1      E  85.00  8.89  1.17                 FORT ERIE - CENTRAL AV IC   
2      W  85.13  8.83  1.08                 FORT ERIE - CENTRAL AV IC   
3      E  80.84  8.18  0.87                            GILMORE RD I/C   
4      W  79.18  8.03  0.91                            GILMORE RD I/C   
5      N  81.24  8.34  1.01               1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                       ...   
1825   B  85.01  8.65  0.88                         WEST JCT OF HWY 2   
1826   B  79.14  8.15  1.06                          ROSE ISLAND LANE   
1827   B  79.49  8.18  1.05                   2.8 KM EAST OF ROCKPORT   
1828   B  72.33  7.91  1.68  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829   B  78.93  8.16  1.10                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To  
1                    FORT ERIE - GILMORE RD IC-5  
2                    FORT ERIE - GILMORE RD IC-5  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
5                 NIAG FALLS - LYONS CK RD IC-21  
...                                          ...  
1825                            ROSE ISLAND LANE  
1826                             NARROWS LANE RD  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  
1828                  0.5 KM W OF MALLORYTOWN RD  
1829                  EAST JUNCTION OF HIGHWAY 2  

[1829 rows x 6 columns]

In [32]:
d2014

DIR    PCI   DMI   IRI                       Pavement_Section_From  \
1      E  85.74  8.97  1.17                 FORT ERIE - GILMORE RD IC-5   
2      W  91.51  9.54  1.08                   FORT ERIE - CENTRAL AV IC   
3      E  89.05  9.09  0.87  NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12   
4      W  88.16  9.03  0.91                              GILMORE RD I/C   
5      N  88.72  9.17  1.01                 1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                         ...   
1814   B  79.81  8.65  1.57                                   ROBINS RD   
1815   B  92.02  9.83  1.36                                      OKE DR   
1816   B  91.95  9.63  1.13                               ONTARIO ST IC   
1817   B  87.60  8.93  0.87                              LAKE FOREST DR   
1818   B  90.34  9.51  1.20                                  HWY 400 IC   

                             Pavement_Section_To   Latitude  Longitude  
1                      FORT ERIE - CENTRAL AV IC  42.919138 -78.968180  
2                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
3                                 GILMORE RD I/C  42.975538 -79.038586  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
5                 NIAG FALLS - LYONS CK RD IC-21  42.975610 -79.038328  
...                                          ...        ...        ...  
1814                                  END OF HWY        NaN        NaN  
1815                                  END OF HWY        NaN        NaN  
1816                                  END OF HWY        NaN        NaN  
1817                                     HWY 559        NaN        NaN  
1818                         HWY 7909 (NOBEL RD)        NaN        NaN  

[1818 rows x 8 columns]

In [33]:
import numpy as np

In [34]:
d2014["Latitude"].isna().sum()

490

In [35]:
#We now compare and match PavementSectionFrom and PavementSectionTo with 2014 dataset and change the latitude longitudes if its not NaN
d2015['Pavement_Section_From'].isin(d2014['Pavement_Section_From']).value_counts()
            
            
        

Pavement_Section_From
True     1767
False      62
Name: count, dtype: int64

In [36]:
import math

In [37]:
#Here we basically check for same from locations in 2015 and 2014 data and check if 2014 latitudes are present, if yes, put those in addressses and later on , we will assign values in addresses to 2015 dataset
for ind in addresses.index:
    for ind1 in d2014.index:
        if addresses['Pavement_Section_From'][ind]==(d2014['Pavement_Section_From'][ind1]) or addresses['Pavement_Section_From'][ind]==(d2014['Pavement_Section_To'][ind1]):
            if math.isnan(d2014['Latitude'][ind1])==False:
                addresses['Lat'][ind]=d2014['Latitude'][ind1]
                addresses['Lon'][ind]=d2014["Longitude"][ind1]
    ind1=0
                
            

C:\Users\NAVEEN\AppData\Local\Temp\ipykernel_2148\2261657229.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  addresses['Lat'][ind]=d2014['Latitude'][ind1]
C:\Users\NAVEEN\AppData\Local\Temp\ipykernel_2148\2261657229.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  addresses['Lon'][ind]=d2014["Longitude"][ind1]
C:\Users\NAVEEN\AppData\Local\Temp\ipykernel_2148\2261657229.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  address

In [38]:
d2014

DIR    PCI   DMI   IRI                       Pavement_Section_From  \
1      E  85.74  8.97  1.17                 FORT ERIE - GILMORE RD IC-5   
2      W  91.51  9.54  1.08                   FORT ERIE - CENTRAL AV IC   
3      E  89.05  9.09  0.87  NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12   
4      W  88.16  9.03  0.91                              GILMORE RD I/C   
5      N  88.72  9.17  1.01                 1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                         ...   
1814   B  79.81  8.65  1.57                                   ROBINS RD   
1815   B  92.02  9.83  1.36                                      OKE DR   
1816   B  91.95  9.63  1.13                               ONTARIO ST IC   
1817   B  87.60  8.93  0.87                              LAKE FOREST DR   
1818   B  90.34  9.51  1.20                                  HWY 400 IC   

                             Pavement_Section_To   Latitude  Longitude  
1                      FORT ERIE - CENTRAL AV IC  42.919138 -78.968180  
2                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
3                                 GILMORE RD I/C  42.975538 -79.038586  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
5                 NIAG FALLS - LYONS CK RD IC-21  42.975610 -79.038328  
...                                          ...        ...        ...  
1814                                  END OF HWY        NaN        NaN  
1815                                  END OF HWY        NaN        NaN  
1816                                  END OF HWY        NaN        NaN  
1817                                     HWY 559        NaN        NaN  
1818                         HWY 7909 (NOBEL RD)        NaN        NaN  

[1818 rows x 8 columns]

In [39]:
addresses

Pavement_Section_From  \
1                    FORT ERIE - CENTRAL AV IC   
2                    FORT ERIE - CENTRAL AV IC   
3                               GILMORE RD I/C   
4                               GILMORE RD I/C   
5                  1.0 KM W OF NETHERBY RD I/C   
...                                        ...   
1825                         WEST JCT OF HWY 2   
1826                          ROSE ISLAND LANE   
1827                   2.8 KM EAST OF ROCKPORT   
1828  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To        Lat        Lon  
1                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
2                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
5                 NIAG FALLS - LYONS CK RD IC-21  43.032582 -79.112933  
...                                          ...        ...        ...  
1825                            ROSE ISLAND LANE  44.251579 -72.571407  
1826                             NARROWS LANE RD  44.353108 -76.039714  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  44.399129 -75.918005  
1828                  0.5 KM W OF MALLORYTOWN RD  44.419990 -75.893940  
1829                  EAST JUNCTION OF HIGHWAY 2  44.452994 -75.860840  

[1827 rows x 4 columns]

In [40]:
addresses['Lat'].isin(d2014['Latitude']).value_counts()

Lat
True     1357
False     470
Name: count, dtype: int64

Now there are lot of common values meaning the latitudes and longitudes have been updated properly, now we can add them to 2015 dataset itself

In [41]:
d2015

DIR    PCI   DMI   IRI                     Pavement_Section_From  \
1      E  85.00  8.89  1.17                 FORT ERIE - CENTRAL AV IC   
2      W  85.13  8.83  1.08                 FORT ERIE - CENTRAL AV IC   
3      E  80.84  8.18  0.87                            GILMORE RD I/C   
4      W  79.18  8.03  0.91                            GILMORE RD I/C   
5      N  81.24  8.34  1.01               1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                       ...   
1825   B  85.01  8.65  0.88                         WEST JCT OF HWY 2   
1826   B  79.14  8.15  1.06                          ROSE ISLAND LANE   
1827   B  79.49  8.18  1.05                   2.8 KM EAST OF ROCKPORT   
1828   B  72.33  7.91  1.68  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829   B  78.93  8.16  1.10                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To  
1                    FORT ERIE - GILMORE RD IC-5  
2                    FORT ERIE - GILMORE RD IC-5  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  
5                 NIAG FALLS - LYONS CK RD IC-21  
...                                          ...  
1825                            ROSE ISLAND LANE  
1826                             NARROWS LANE RD  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  
1828                  0.5 KM W OF MALLORYTOWN RD  
1829                  EAST JUNCTION OF HIGHWAY 2  

[1829 rows x 6 columns]

In [42]:
addresses

Pavement_Section_From  \
1                    FORT ERIE - CENTRAL AV IC   
2                    FORT ERIE - CENTRAL AV IC   
3                               GILMORE RD I/C   
4                               GILMORE RD I/C   
5                  1.0 KM W OF NETHERBY RD I/C   
...                                        ...   
1825                         WEST JCT OF HWY 2   
1826                          ROSE ISLAND LANE   
1827                   2.8 KM EAST OF ROCKPORT   
1828  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To        Lat        Lon  
1                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
2                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
5                 NIAG FALLS - LYONS CK RD IC-21  43.032582 -79.112933  
...                                          ...        ...        ...  
1825                            ROSE ISLAND LANE  44.251579 -72.571407  
1826                             NARROWS LANE RD  44.353108 -76.039714  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  44.399129 -75.918005  
1828                  0.5 KM W OF MALLORYTOWN RD  44.419990 -75.893940  
1829                  EAST JUNCTION OF HIGHWAY 2  44.452994 -75.860840  

[1827 rows x 4 columns]

In [43]:
#Joining them
d2015_final=d2015.copy()

In [44]:
d2015_final['Lat']=addresses['Lat']
d2015_final['Lon']=addresses['Lon']

In [45]:
addresses

Pavement_Section_From  \
1                    FORT ERIE - CENTRAL AV IC   
2                    FORT ERIE - CENTRAL AV IC   
3                               GILMORE RD I/C   
4                               GILMORE RD I/C   
5                  1.0 KM W OF NETHERBY RD I/C   
...                                        ...   
1825                         WEST JCT OF HWY 2   
1826                          ROSE ISLAND LANE   
1827                   2.8 KM EAST OF ROCKPORT   
1828  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To        Lat        Lon  
1                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
2                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
5                 NIAG FALLS - LYONS CK RD IC-21  43.032582 -79.112933  
...                                          ...        ...        ...  
1825                            ROSE ISLAND LANE  44.251579 -72.571407  
1826                             NARROWS LANE RD  44.353108 -76.039714  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  44.399129 -75.918005  
1828                  0.5 KM W OF MALLORYTOWN RD  44.419990 -75.893940  
1829                  EAST JUNCTION OF HIGHWAY 2  44.452994 -75.860840  

[1827 rows x 4 columns]

In [46]:
d2015_final

DIR    PCI   DMI   IRI                     Pavement_Section_From  \
1      E  85.00  8.89  1.17                 FORT ERIE - CENTRAL AV IC   
2      W  85.13  8.83  1.08                 FORT ERIE - CENTRAL AV IC   
3      E  80.84  8.18  0.87                            GILMORE RD I/C   
4      W  79.18  8.03  0.91                            GILMORE RD I/C   
5      N  81.24  8.34  1.01               1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                       ...   
1825   B  85.01  8.65  0.88                         WEST JCT OF HWY 2   
1826   B  79.14  8.15  1.06                          ROSE ISLAND LANE   
1827   B  79.49  8.18  1.05                   2.8 KM EAST OF ROCKPORT   
1828   B  72.33  7.91  1.68  5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)   
1829   B  78.93  8.16  1.10                0.5 KM W OF MALLORYTOWN RD   

                             Pavement_Section_To        Lat        Lon  
1                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
2                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
3     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
5                 NIAG FALLS - LYONS CK RD IC-21  43.032582 -79.112933  
...                                          ...        ...        ...  
1825                            ROSE ISLAND LANE  44.251579 -72.571407  
1826                             NARROWS LANE RD  44.353108 -76.039714  
1827    5.7 KM EAST OF ROCKPORT(POOLS RESORT RD)  44.399129 -75.918005  
1828                  0.5 KM W OF MALLORYTOWN RD  44.419990 -75.893940  
1829                  EAST JUNCTION OF HIGHWAY 2  44.452994 -75.860840  

[1829 rows x 8 columns]

In [47]:
#d2015_final is ready to be used to fetch weather data, now similarly, lets preprocess 2014,2015,2016,2017 data as well
d2014

DIR    PCI   DMI   IRI                       Pavement_Section_From  \
1      E  85.74  8.97  1.17                 FORT ERIE - GILMORE RD IC-5   
2      W  91.51  9.54  1.08                   FORT ERIE - CENTRAL AV IC   
3      E  89.05  9.09  0.87  NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12   
4      W  88.16  9.03  0.91                              GILMORE RD I/C   
5      N  88.72  9.17  1.01                 1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                         ...   
1814   B  79.81  8.65  1.57                                   ROBINS RD   
1815   B  92.02  9.83  1.36                                      OKE DR   
1816   B  91.95  9.63  1.13                               ONTARIO ST IC   
1817   B  87.60  8.93  0.87                              LAKE FOREST DR   
1818   B  90.34  9.51  1.20                                  HWY 400 IC   

                             Pavement_Section_To   Latitude  Longitude  
1                      FORT ERIE - CENTRAL AV IC  42.919138 -78.968180  
2                    FORT ERIE - GILMORE RD IC-5  42.909536 -78.918073  
3                                 GILMORE RD I/C  42.975538 -79.038586  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210 -78.967961  
5                 NIAG FALLS - LYONS CK RD IC-21  42.975610 -79.038328  
...                                          ...        ...        ...  
1814                                  END OF HWY        NaN        NaN  
1815                                  END OF HWY        NaN        NaN  
1816                                  END OF HWY        NaN        NaN  
1817                                     HWY 559        NaN        NaN  
1818                         HWY 7909 (NOBEL RD)        NaN        NaN  

[1818 rows x 8 columns]

Some latitude values are missing, so we add those

In [48]:

for ind in d2014.index:
    if math.isnan(d2014['Latitude'][ind]):
        hh=latlong(d2014['Pavement_Section_From'][ind])
        d2014["Latitude"][ind],d2014['Longitude'][ind]=hh

C:\Users\NAVEEN\AppData\Local\Temp\ipykernel_2148\3334264987.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  d2014["Latitude"][ind],d2014['Longitude'][ind]=hh
C:\Users\NAVEEN\AppData\Local\Temp\ipykernel_2148\3334264987.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  d2014["Latitude"][ind],d2014['Longitude'][ind]=hh
C:\Users\NAVEEN\AppData\Local\Temp\ipykernel_2148\3334264987.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  

In [49]:
d2014['Longitude'].value_counts()

Longitude
-86.836052    11
-74.033676     9
-80.769539     6
-0.650476      6
 74.501518     4
              ..
-79.175431     1
-80.792384     1
-80.752946     1
-94.941713     1
-77.045936     1
Name: count, Length: 1754, dtype: int64

In [50]:
d2014=d2014[d2014['Longitude']!=0]

In [51]:
d2015_final['Lon'].isin(d2014['Longitude']).value_counts()

Lon
True     1787
False      42
Name: count, dtype: int64

In [52]:
d2014

DIR    PCI   DMI   IRI                       Pavement_Section_From  \
1      E  85.74  8.97  1.17                 FORT ERIE - GILMORE RD IC-5   
2      W  91.51  9.54  1.08                   FORT ERIE - CENTRAL AV IC   
3      E  89.05  9.09  0.87  NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12   
4      W  88.16  9.03  0.91                              GILMORE RD I/C   
5      N  88.72  9.17  1.01                 1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                         ...   
1814   B  79.81  8.65  1.57                                   ROBINS RD   
1815   B  92.02  9.83  1.36                                      OKE DR   
1816   B  91.95  9.63  1.13                               ONTARIO ST IC   
1817   B  87.60  8.93  0.87                              LAKE FOREST DR   
1818   B  90.34  9.51  1.20                                  HWY 400 IC   

                             Pavement_Section_To   Latitude   Longitude  
1                      FORT ERIE - CENTRAL AV IC  42.919138  -78.968180  
2                    FORT ERIE - GILMORE RD IC-5  42.909536  -78.918073  
3                                 GILMORE RD I/C  42.975538  -79.038586  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210  -78.967961  
5                 NIAG FALLS - LYONS CK RD IC-21  42.975610  -79.038328  
...                                          ...        ...         ...  
1814                                  END OF HWY  42.048074  -91.672911  
1815                                  END OF HWY  45.636451  -79.413469  
1816                                  END OF HWY  41.799060  -71.422014  
1817                                     HWY 559  33.629013 -117.717311  
1818                         HWY 7909 (NOBEL RD)  38.810012  -77.045936  

[1817 rows x 8 columns]

In [56]:
d2014

DIR    PCI   DMI   IRI                       Pavement_Section_From  \
1      E  85.74  8.97  1.17                 FORT ERIE - GILMORE RD IC-5   
2      W  91.51  9.54  1.08                   FORT ERIE - CENTRAL AV IC   
3      E  89.05  9.09  0.87  NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12   
4      W  88.16  9.03  0.91                              GILMORE RD I/C   
5      N  88.72  9.17  1.01                 1.0 KM W OF NETHERBY RD I/C   
...   ..    ...   ...   ...                                         ...   
1814   B  79.81  8.65  1.57                                   ROBINS RD   
1815   B  92.02  9.83  1.36                                      OKE DR   
1816   B  91.95  9.63  1.13                               ONTARIO ST IC   
1817   B  87.60  8.93  0.87                              LAKE FOREST DR   
1818   B  90.34  9.51  1.20                                  HWY 400 IC   

                             Pavement_Section_To   Latitude   Longitude  
1                      FORT ERIE - CENTRAL AV IC  42.919138  -78.968180  
2                    FORT ERIE - GILMORE RD IC-5  42.909536  -78.918073  
3                                 GILMORE RD I/C  42.975538  -79.038586  
4     NIAG FALLS - 1.0 KM W OF NETHERBY RD IC-12  42.919210  -78.967961  
5                 NIAG FALLS - LYONS CK RD IC-21  42.975610  -79.038328  
...                                          ...        ...         ...  
1814                                  END OF HWY  42.048074  -91.672911  
1815                                  END OF HWY  45.636451  -79.413469  
1816                                  END OF HWY  41.799060  -71.422014  
1817                                     HWY 559  33.629013 -117.717311  
1818                         HWY 7909 (NOBEL RD)  38.810012  -77.045936  

[1817 rows x 8 columns]

In [68]:
from pathlib import Path

In [70]:
filepath=Path('G:\ml notebooks\Road_RUL\datasets\datasets\d2014.csv')
filepath1=Path('G:\ml notebooks\Road_RUL\datasets\datasets\d2015_final.csv')
filepath.parent.mkdir(parents=True,exist_ok=True)
filepath1.parent.mkdir(parents=True,exist_ok=True)


In [71]:
d2014.to_csv(filepath)
d2015_final.to_csv(filepath1)

Now lets focus on getting Coordinates for 2016 and 2017 as well which we will do in another notebook.